Add configurable delay between connect and first request in modbus

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Some devices like Huawei sun2000 4ktl need a pause after the connection is established before receiving requests.

This had lead to instability for users with such devices, they experience HA starts, modbus sensors etc are updated, and
then the platforms becomes unavailable (device disconnected, and modbus reconnects) this pattern can continue a while, until the device accept the first request. In some bad situation users have experienced slow responses from the frontend, this PR might help with that, because there should not be regular disconnects at startup.

with:
```
modbus:
     ....
     delay: 10
     ...
```
there will be a pause of 10 seconds (+ scan_interval)  between the connect() and the first request.

If the write service is called or e.g. turn_on/turn_off on a platform, these call will return False and not be executed 
within the delay period.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
The schema was already in place and have not been changed, nor the documentation. This PR only activate the delay internally.

- This PR fixes or closes issue: fixes #
fixes 50011
- This PR is related to issue: 
- Link to documentation pull request: 
https://github.com/home-assistant/home-assistant.io/pull/17703

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
Activate startup delay.

janiversen commented:
@vzahradnik if you have time, it would be nice if you could try this on your systems as I have done on my system. And of please do a review.

vzahradnik commented:
Ok, I'll check tomorrow.

MartinHjelmare commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -150,10 +151,6 @@ def __init__(self, client_config):
-            self._config_delay = client_config[CONF_DELAY]
-
-        if self._config_delay > 0:
-            _LOGGER.warning("Parameter delay is accepted but not used in this version")
```
Was this an old feature that we had deprecated?

janiversen replied:
yes a leftover from when we tried async pymodbus.

MartinHjelmare commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -208,6 +205,14 @@ def setup(self):
 
+        # Start counting down to allow modbus requests.
+        if self._config_delay:
+            async_call_later(hass, self._config_delay, self.end_delay)
```
Do we need to unsubscribe this listener on Home Assistant stop event?

janiversen replied:
Good question, if HA starts and are stopped within a couple of seconds then yes.

I am not sure how to unsubscribe a "call_later", but I will check for a call to do that.

MartinHjelmare replied:
Store the return value here and call it to unsubscribe.

janiversen replied:
That was easy done as a security measure.

MartinHjelmare commented on tests/components/modbus/test_init.py:
```diff
@@ -467,3 +474,116 @@ async def test_pymodbus_connect_fail(hass, caplog, mock_pymodbus):
+    await hass.services.async_call(DOMAIN, SERVICE_WRITE_REGISTER, data, blocking=True)
+    assert not mock_pymodbus.write_registers.called
+
+    # 2 times fire_changed is needed to secure "normal" update is called.
```
I don't follow why we need to wait 2 x 30 seconds. Hasn't the 15 second delay already expired within the first 30 seconds?

janiversen replied:
It might very well be that the interval can be set a lot shorter. What I saw in test was that (for a reason I do not understand) 2 fire_changed time was needed. I suppose it has something to do with how the event queue are created and emptied.

While I do the change above, I will play with smaller delays. However it does not make the test faster :-)

janiversen replied:
Reduced to 6seconds and 10seconds, scraping it lower might be possible, but I experienced that repeated run of the test would sometime break.

MartinHjelmare replied:
I think I understand what is happening now.

Time doesn't actually pass in the test. We just fire the fake time changed event. The first time changed event where time is more than the delay will call the callback to reset the delay. Then another time changed event is required to actually update the state of the entities.

Does that make sense? That would mean that the new times don't update the state reliably but depends on the order of tasks.

janiversen replied:
your understanding is correct, I suspect it depends on the order of the events (call_later and the scheduled update, but admitted I did not debug the core event handler.

I have seen this earlier in other tests too, if you have 2 listeners with events in the same timeslot the sequence can vary.

janiversen committed:
Add removal of call_later if HA is stopped.

This is unlikely to happen, but just security measure.

janiversen committed:
Removing timing interval.

async_fire_time_changed() needs to be called twice, first time the delay is
ended and second time update() is executed.

janiversen committed:
Variable naming.

vzahradnik commented:
> Ok, I'll check tomorrow.

Plans change :) I will try to test it on Friday.

vzahradnik reviewed (approved):
I tested your change with my setup. It looks OK for me, too.
